lib/graphql/batch/setup.rb

```diff
@@ -1,14 +1,54 @@
 module GraphQL::Batch
-  module Setup
-    extend self
+  class Setup
+    class << self
+      def start_batching
+        raise NestedError if GraphQL::Batch::Executor.current
+        GraphQL::Batch::Executor.current = GraphQL::Batch::Executor.new
+      end
+
+      def end_batching
+        GraphQL::Batch::Executor.current = nil
+      end
+
+      def instrument_field(schema, type, field)
+        return field unless type == schema.mutation
+        old_resolve_proc = field.resolve_proc
+        field.redefine do
+          resolve ->(obj, args, ctx) {
+            GraphQL::Batch::Executor.current.clear
+            begin
+              Promise.sync(old_resolve_proc.call(obj, args, ctx))
```

Comment:
If a loader were re-used in a mutation to load data to mutate, then it wouldn't be necessary to call sync on the promise.  E.g. a callback chain like `RecordLoader.for(Model).load(args[:id]).then { ... }` could be used.  That isn't a pattern that we use, but it doesn't make sense to prevent that considering graphql-ruby provides lazy loading support for mutation fields.  So rather than change that behaviour, I just want to try to make sure the cache is actually cleared after any promise callbacks on a returned callback have been called.